RFCs/rfc-0007-zkPassport.md

```diff
@@ -0,0 +1,150 @@
+
+Using this information directly online would expose passport information online, which would not be viable.
+
+However, with ZKPs, users could provide proofs about their passport information, without exposing the passport information itself.
+
+Additionally, because passports following this standard support NFC, which is present on most smartphones, users can obtain this data themselves from their own passports relatively easily. Here for example is an open library that reads passports for iOS[[3](https://github.com/AndyQ/NFCPassportReader)].
+
+Properties stored in the standard include:
+* documentType
+* documentSubType
+* documentNumber
+* issuingAuthority
+* documentExpiryDate
+* dateOfBirth
+* gender
+* nationality
+* lastName
+* firstName
+* names
+* placeOfBirth
+* residenceAddress
+* phoneNumber
+
+## Objectives
+
+We propose a grant taking advantage of this, implementing, at a high level, the following:
+
+1. An iOS & Android App that uses NFC to allow users to get the data from their own passports, and store that data in their MINA wallet for secure use by web3 applications through the (in-progress) Mina Attestation API Standard.
+2. An o1js library that uses the Mina Attestation API Standard to create proofs of various components of one's zkPassport identity, for use in proof-of-unique-user and proof of nationality & age use cases.
+
+## Motivation and Rationale
+
+This would unlock many use cases for web3 and web2 applications connected to Mina.
+
+For applications on Mina, they can directly take advantage of this data through nullifiers for proof-of-uniqueness and other identity related proofs (see both scenarios below).
+
+For other chains, using Mina's zk state bridge, they could also leverage these proofs in their applications.
+
+Web2 applications may find use in this functionality as well. This seems particularly relevant as AI systems become increasingly powerful, and the risk this could pose to new and existing digital systems being overwhelmed by fake users.
+
+## Scenarios and Use Cases
+
+### Example Scenario 1: Proof-of-Uniqueness
+
+| Aspect           | Description |
+|------------------|-------------|
+| **Description**  | A zkApp wants a user to prove they are unique. |
+| **Requirements** | An implementation of zkPassport in the Attestation API Standard (see objectives), plus an on-chain nullifier that users can use with the zkApp. zkPassport nullifier proofs should be sure to have a *context* input salt that is hashed with the unique zkPassport identifier to ensure identities are unique across their context, but cannot be associated cross-context unless a proof is provided by the user. |
```

Comment:
@KimlikDAO-bot this is one way to ensure that identities aren't traceable cross application - the salt can make sure that how one proves uniqueness for one application isn't traceable across applications.

It would probably be important that however wallets interact with nullifiers, wallets check and make sure the same nullifier-key isn't being used across applications

Reply:
Yeah that's a good point. I think for this passport case, I think there should be enough data to avoid this, by using the passport signature from the authority itself as part of the hash - so `unique_passport_hash = hash(passportSignature, passportData)`, and it shouldn't be brute-forceable I think.

And then using `hash(application_salt, unique_passport_hash)` for any nullifiers on-chain to avoid revealing the unique_passport_hash, seems like it should work. Even if the application_salt is constant, there shouldn't be a way to find the unique_passport_hash then, (or the additional requirement of creating a proof that one possesses the data underlying that hash that has been signed by the passport authority).

Any group implementing / auditors should definitely make sure this works out in the implementation and specifics of any passport cryptography though, as well as put together a more complete spec. Curious if you think this seems right or any other thoughts though.

Reply:
Ah I see, yeah that makes a lot of sense. Yeah it would probably be best to avoid the passport authority being able to do that.

Maybe the following could be done to do this without needing a multiparty protocol? (and any additional trust assumptions that would need to follow from that?)

1) have an "identity-registration" zkApp, that uses a nullifier using the passport-authority-visible hash as the key, to allow users to register their own user-selected unique hash.
2) have other applications use the user-selected hash for proofs of uniqueness - which should work because each user-selected hash corresponds to an passport-authority-generated hash, and there isn't a way to generate multiple user-selected hashes per passport-authority generated credential.

Curious on any thoughts? Seems like something multistage is needed though yeah for on-chain privacy from the passport authority.

Reply:
Interesting re the encryption size thing - I'd think that if its big enough to not be brute-forceable to fake-sign passports, it should be enough entropy for not brute-forcing a hash of all of the data? Though I haven't looked at the cryptography in detail, and would definitely be curious for a cryptographers take.

Reply:
The encryption is there to protect the passport data but majority of implementations are very week at that level. The certificate signed by issuing government is mostly safe. There are few mistakes done by few governments in the history while transitioning towards the e-passports but I don't expect similar incidents being a issue in the future.

Reply:
I'm not sure, it should be sufficient I think to get this property (on-chain privacy from the issuer) through an on-chain operation? Could totally be missing something though.

In particular:
1) Alice generates S, their private, unique identifier, and stores it off-chain
2) Alice receives P, their unique passport credential, and H(P), the hash of P, which is private to Alice and the passport-issuing-authority.
3) An on-chain identity zkApp is created with a nullifier for registering identities, which accepts a proof, key, and value, and application salt A.
4) Alice posts to the on-chain identity zkApp:
    a) a proof that Alice posseses P signed by the passport issuing authority
    b) a key into the nullifier, H(A,H(P))
    c)  a value to register into the nullifier, H(S)
5) Afterwards, Alice proves to other applications that she knows of an S that has an H(S) in the on-chain identity zkApp nullifier, and when using other applications, nullifiers, uses the key H(Application_key), S).

Because S is private to Alice, nobody else can know which passport-authority-issued identity her identifier S is tied to, or her private value S, so it should be resolve potential privacy issues with the passport-issuer, but lmk, definitely could have missed something.

Reply:
I believe this is mostly inline with what we were discussing, with @RaidasGrisk.

Reply:
@es92 if the key is a Merkle map key (that's how I understood your idea), then you have to share that key as you update the onchain Merkle root. Otherwise, others can not keep track of the updated Merkle tree

Reply:
What would be nice is some sort of non inclusion proof into a set without revealing anything else

Reply:
Maybe something like a bloom filter, where the hash of your unique secret is smashed together with others and so by proving that your set addition toggles 1 bit, you only reveal a tiny amount

Reply:
Actually, what's so bad about revealing a passport hash at the time of registration in that zkApp. You can do that from any account. If you later prove inclusion you don't reveal anything. I.e. I think @es92 your solution is good.

Also, why does it have to be an application specific hash? Why not take H(P) as the Merkle map key?

Reply:
@KimlikDAO-bot I understand that you can brute-force `P`, i.e. the users personal data, from `H(P)`.

However, as I've tried to convey, this shouldn't be a problem, because you wouldn't use `H(P)` as your "ID". Instead, you would only use `H(P)` in a registration step, to register a _different_ ID, `H(S)` - let's say, for simplicity, a hash of your private key.

After that, you would be able to create "proofs of unique personhood" _without revealing anything about your actual identity_: You demonstrate, in a zk proof, that you know _private inputs_ `S` (private key) and a `P` (passport data) such that `H(S)` is included in a Merkle tree, at position `H(P)`. This zk computation would just be part of a zkApp which requires your unique-humanness.

Reply:
So in a sense there is no "number that is your unique id". That is not needed. What's needed is a way to zk-prove that you _have_ a unique id

Reply:
> Anyone who has read the NFC chip of your passport (airlines, airport authorities) or anyone who can recreate that (government signing authorities) can immediately link your wallet address to your government ID

They can link the _address that you register your id from_ to your personal data. However, your later actions are unlinkable with this step.

This is very similar to the kind of anonymity that protocols like ZCash or Tornado Cash give you: There is a step where you send money into a shielded pool, from a public address. So, everyone can see that this address sends money into the pool. In our system, everyone can see that some person registers for a unique id.

But after registering, your actions are protected by the anonymity set of all the people who also registered.

Reply:
However, it feels like front-running is a dangerous issue in that design @es92
Someone who sees that transaction registering `H(P)` into the Merkle map can just register his own secret to the same `H(P)` first, thus stealing your passport identity 😨

Reply:
Well actually I think the front-running thing can be solved with a commit-and-reveal scheme. You "preregister" first, by showing a blinding commitment to `P` i.e. `H(S, P)`. And it's only possible to register once n blocks have passed since a preregister was done. So, unless someone manages to prevent you from posting a transaction completely for n blocks, they aren't able to front-run it because they need to create their own preregistration first

Reply:
> Well actually I think the front-running thing can be solved with a commit-and-reveal scheme. You "preregister" first, by showing a blinding commitment to `P` i.e. `H(S, P)`. And it's only possible to register once n blocks have passed since a preregister was done. So, unless someone manages to completely block you from posting a transaction for n blocks, they aren't able to front-run it because they'd need to create their own preregistration _after_ seeing your data

Yep - further, the zkApp can require a proof that alice knows S in H(S), so knowing H(S) shouldn't be enough to front-run then either (if we're talking about front-running H(S)).

I'm not sure on the entropy, but I'll add that to a "todo" section of this RFC - I'd think there'd be enough entropy between all the data in P (includes I think maybe fingerprint / photo data as well?) and the signature from the issuing authority, but its definitely something someone on the cryptography side should check (it sounds like definitely a risk at least to make sure of).

Reply:
Depends on passport since they are all around the place but majority of them have very low entropy because of limited bit space they have and data they need to put in to that space.

Reply:
https://www.google.com/url?sa=t&source=web&rct=j&opi=89978449&url=https://eprint.iacr.org/2005/095.pdf&ved=2ahUKEwiBx5DusKOEAxVO9gIHHUo8AaYQFnoECDYQAQ&usg=AOvVaw2LKSzfBduBGPNOm2sFL_7F

This is the most detailed document. While this is old if you do your research you can find this still applies to many passports globally. The biggest problem comes from general practices since this paper didn't change. Few countries changed their structure and it's safer, majority of them didn't make that. There is a new standard published but it's a big question when it's adopted or even at all.